storage/src/test/java/tech/pegasys/teku/storage/client/RecentChainDataTest.java

```diff
@@ -854,6 +877,86 @@ public void getAncestorsOnFork_unknownRoot() {
     assertThat(recentChainData.getAncestorsOnFork(UInt64.valueOf(1), Bytes32.ZERO)).isEmpty();
   }
 
+  @Test
+  void isOptimisticSyncPossible_shouldBeFalseWhenMergeBlockDoesNotExistAndBlockTooRecent() {
+    initPostGenesis();
+    assertThat(recentChainData.isOptimisticSyncPossible(recentChainData.getHeadSlot())).isFalse();
+  }
+
+  @Test
+  void isOptimisticSyncPossible_shouldBeFalseWhenMergeBlockNotJustifiedAndBlockTooRecent() {
+    initPostGenesis();
+    storageSystem
+        .chainUpdater()
+        .saveBlock(
+            chainBuilder.generateBlockAtSlot(
+                recentChainData.getHeadSlot().plus(1),
+                BlockOptions.create().setTerminalBlockHash(dataStructureUtil.randomBytes32())));
+    assertThat(recentChainData.isOptimisticSyncPossible(recentChainData.getHeadSlot())).isFalse();
+  }
+
+  @Test
+  void isOptimisticSyncPossible_shouldBeTrueWhenGenesisIsPostMerge() {
+    initPreGenesis();
+    storageSystem.chainUpdater().initializeGenesisWithPayload(false);
+    assertThat(recentChainData.isOptimisticSyncPossible(recentChainData.getHeadSlot())).isTrue();
+  }
+
+  @Test
+  void isOptimisticSyncPossible_shouldBeTrueWhenMergeBlockNotJustifiedAndBlockOldEnough() {
+    initPostGenesis();
+    final int safeSyncDistance =
+        spec.getGenesisSpecConfig()
+            .toVersionMerge()
+            .orElseThrow()
+            .getSafeSlotsToImportOptimistically();
+    final UInt64 mergeBlockSlot = recentChainData.getHeadSlot().plus(1);
+    final UInt64 importBlockSlot = mergeBlockSlot.plus(1);
+    storageSystem
+        .chainUpdater()
+        .saveBlock(
+            chainBuilder.generateBlockAtSlot(
+                mergeBlockSlot,
+                BlockOptions.create().setTerminalBlockHash(dataStructureUtil.randomBytes32())));
+
+    storageSystem.chainUpdater().setCurrentSlot(importBlockSlot.plus(safeSyncDistance));
+    assertThat(recentChainData.isOptimisticSyncPossible(importBlockSlot)).isTrue();
+  }
+
+  @Test
+  void isOptimisticSyncPossible_shouldBeFalseWhenMergeBlockIsOldButImportingBlockIsRecent() {
+    initPostGenesis();
+    final UInt64 safeSyncDistance = UInt64.valueOf(128);
```

Comment:
is intended to use two different ways to get `safeSyncDistance`? here: hardcoded, in previous test: from spec.

Reply:
Yeah that should have come from the spec too, good spot.